ee/middleware/upstreamoauth/provider.go

```diff
@@ -0,0 +1,300 @@
+	HeaderName string
+	// AuthValue is the value of auth header.
+	AuthValue string
+}
+
+// Fill sets the request's HeaderName with AuthValue
+func (u Provider) Fill(r *http.Request) {
+	if r.Header.Get(u.HeaderName) != "" {
+		u.Logger.WithFields(logrus.Fields{
+			"header": u.HeaderName,
+		}).Info("Authorization header conflict detected: Client header overwritten by Gateway upstream authentication header.")
+	}
+	r.Header.Set(u.HeaderName, u.AuthValue)
+}
+
+type OAuthHeaderProvider interface {
+	// getOAuthToken returns the OAuth token for the request.
+	getOAuthToken(r *http.Request, OAuthSpec *Middleware) (string, error)
+	// getHeaderName returns the header name for the OAuth token.
+	getHeaderName(OAuthSpec *Middleware) string
+	//
+	headerEnabled(OAuthSpec *Middleware) bool
+}
+
+func getOAuthHeaderProvider(oauthConfig apidef.UpstreamOAuth) (OAuthHeaderProvider, error) {
+	if !oauthConfig.IsEnabled() {
+		return nil, fmt.Errorf("upstream OAuth is not enabled")
+	}
+
+	switch {
+	case len(oauthConfig.AllowedAuthorizeTypes) == 0:
+		return nil, fmt.Errorf("no OAuth configuration selected")
+	case len(oauthConfig.AllowedAuthorizeTypes) > 1:
+		return nil, fmt.Errorf("both client credentials and password authentication are provided")
+	case oauthConfig.AllowedAuthorizeTypes[0] == ClientCredentialsAuthorizeType:
+		return &ClientCredentialsOAuthProvider{}, nil
+	case oauthConfig.AllowedAuthorizeTypes[0] == PasswordAuthorizeType:
+		return &PasswordOAuthProvider{}, nil
+	default:
+		return nil, fmt.Errorf("no valid OAuth configuration provided")
+	}
+}
+
+func (p *ClientCredentialsOAuthProvider) getOAuthToken(r *http.Request, mw *Middleware) (string, error) {
+	client := UpstreamOAuthClientCredentialsClient{mw.clientCredentialsStorageHandler}
+	token, err := client.GetToken(r, mw)
+	if err != nil {
+		return handleOAuthError(r, mw, err)
+	}
+
+	return fmt.Sprintf("Bearer %s", token), nil
+}
+
+func handleOAuthError(r *http.Request, mw *Middleware, err error) (string, error) {
+	mw.EmitUpstreamOAuthEvent(r, ErrorEventName, err.Error(), mw.Spec.APIID)
+	return "", err
+}
+
+func (p *ClientCredentialsOAuthProvider) getHeaderName(OAuthSpec *Middleware) string {
+	return OAuthSpec.Spec.UpstreamAuth.OAuth.ClientCredentials.Header.Name
+}
+
+func (p *ClientCredentialsOAuthProvider) headerEnabled(OAuthSpec *Middleware) bool {
+	return OAuthSpec.Spec.UpstreamAuth.OAuth.ClientCredentials.Header.Enabled
+}
+
+type UpstreamOAuthCache interface {
+	// GetToken returns the token from cache or issues a request to obtain it from the OAuth provider.
+	GetToken(r *http.Request, OAuthSpec *Middleware) (string, error)
+	// ObtainToken issues a request to obtain the token from the OAuth provider.
+	ObtainToken(ctx context.Context, OAuthSpec *Middleware) (*oauth2.Token, error)
+}
+
+func (cache *UpstreamOAuthClientCredentialsClient) GetToken(r *http.Request, OAuthSpec *Middleware) (string, error) {
```

Comment:
Consider the package name, it's stuttering with all the types, can this be...

upstreamoauth.ClientCredentialsAuth?
upstreamoauth.PasswordAuth?